syntax = "proto3";

package nestbox.protos;

option csharp_namespace = "Nestbox.Protos";  // This line tells protoc the namespace to use for C#


enum Dimension {
    UNKNOWN = 0;  // Default value, representing an unspecified dimension
    X = 1;
    Y = 2;
    Z = 3;
    T = 4;        // Time
    VX = 5;       // Velocity in X
    VY = 6;       // Velocity in Y
    VZ = 7;       // Velocity in Z
}


message CovarianceMatrix {
    repeated float upper_triangle = 1;  // Include diagonal and upper triangle elements, row-major order
}

message Sample {
    repeated float mean = 1;
    CovarianceMatrix covariance = 2;
}

message TransformationMatrix {
    repeated float data = 1;  // Flattened matrix data, row-major order
}

message MeasurementSet {
    repeated Dimension dimensions = 1;  // List of dimensions that this measurement regards
    repeated Sample samples = 2;
    TransformationMatrix transform = 3;
    bool isHomogeneous = 4;  // Indicates if the measurement is in homogeneous coordinates
}

message Twig {
    string coordSysId = 1;  // Unique identifier for the coordinate system
    string streamId = 2;  // Unique identifier for a stream of related measurements
    repeated MeasurementSet measurements = 3;
}
